rfcs/0100-sign-commits.md

```diff
@@ -0,0 +1,345 @@
+---
+feature: sign-commits
+start-date: 2021-08-14
+author: Las Safin
+co-authors:
+shepherd-team:
+shepherd-leader:
+related-issues: https://github.com/NixOS/rfcs/pull/34
+---
+
+# Summary
+[summary]: #summary
+
+- Add a way for flake inputs that point to Git repositories to be verified for trust.
+- Speed up fetching Git repositories with Nix.
+- Recommend and eventually require that people who commit directly to
+  https://github.com/NixOS/nixpkgs and https://github.com/NixOS/nix,
+  so called authorized committers, sign their commits.
+  - Commits inside merges do not have to be signed as long as the merge commit itself is signed,
+		such that only Committers have to sign their commits.
+
+# Motivation
+[motivation]: #motivation
+
+Signing commits would allow cryptographically verifying that the source code
+hasn't been tampered by an untrusted party.
+
+This would ensure that when pulling from GitHub, or some other source,
+you can check if the source code has been signed by a trusted person.
+
+This is already done by Guix:
+- https://guix.gnu.org/en/blog/2020/securing-updates/
+- https://github.com/guix-mirror/guix/commits/master
+- https://guix.gnu.org/manual/en/html_node/Commit-Access.html
+- https://guix.gnu.org/manual/en/html_node/Channel-Authentication.html
+
+It is also done by most conventional distros, however, they sign
+the output from the compilation rather than the source code, since
+they distribute the output from the compilation.
+
+This change would also allow using 3rd-party mirrors of Nixpkgs,
+since we can verify the source code independent of where it came from.
+
+This RFC was motivated by work during Summer of Nix, specifically,
+the Libre SoC maintainer requested that source integrity be
+maintained.
+
+# Detailed design
+[design]: #detailed-design
+
+## Signing the commits
+[signing]: #signing-the-commits
+
+The following will be required from 2022-03-01T00:00Z on,
+but is strongly recommended from now on.
+
+Currently merging is done through GitHub, which would not work in this case.
+Merging would have to happen outside of GitHub, so that you can sign the commits,
+since currently the API doesn't support signing commits with your own PGP
+key: https://github.com/cli/cli/issues/1318.
+
+We would make a script for merging pull requests instead, something like:
```

Comment:
Thanks for the very good suggestions. This kind of tooling would be useful for more than just Nixpkgs. I think we could merge this functionality into GitHub's CLI and Desktop tooling, though preferably, we'd have something that isn't specific to GitHub, since this functionality is useful for more than just Nixpkgs.

Reply:
A lot of this is GPG-specific, which we want to avoid in the first place. Sequoia has a much better UX, so it's not something we need to worry about if users begin using the git-gpg-shim referenced in the RFC.

The last one is a problem with the primitive script I've made, but since the goal will be to integrate the functionality into GitHub Desktop, etc. it will be solved as a part of that.